crates/teloxide/examples/composite_state.rs

```diff
@@ -0,0 +1,190 @@
+//! Imagine that your dialogue state is logically represented by separate
+//! stages, say "setup stage", "perform action stage", etc. Instead of inflating
+//! a single-state enumeration like this:
+//! ```
+//! #[derive(Clone, Default)]
+//! pub enum State {
+//!     #[default]
+//!     Unconfigured,
+//!     ReceiveFullName,
+//!     ReceiveAge {
+//!         full_name: String,
+//!     },
+//!     // Many more variants...
+//!     Idle,
+//! }
+//! ```
+//!
+//! The more appropriate way is to nest enumerations like this:
+//! ```
+//! #[derive(Clone, Default)]
+//! pub enum GlobalState {
+//!     #[default]
+//!     Unconfigured,
+//!     UserSetup(UserSetup),
+//!     // Many more complex stages...
+//!     Idle,
+//! }
+//!
+//! #[derive(Clone)]
+//! enum UserSetup {
+//!     ReceiveFullName,
+//!     ReceiveAge { full_name: String },
+//! }
+//!
+//! // More enumeration definitions...
+//! ```
+//!
+//! This example demonstrates how to achieve this `teloxide` design pattern.
+
+use teloxide::{
+    dispatching::{dialogue::InMemStorage, MessageFilterExt},
+    prelude::*,
+    types::{ChatId, Message},
+};
+
+type Bot = teloxide::Bot;
+type Error = Box<dyn std::error::Error + Send + Sync>;
+type HandlerResult = Result<(), Error>;
+type UpdateHandler = teloxide::dispatching::UpdateHandler<Error>;
+type Storage = InMemStorage<GlobalState>;
+type Dialogue = teloxide::dispatching::dialogue::Dialogue<GlobalState, Storage>;
+
+#[derive(BotCommands, Clone)]
+#[command(rename_rule = "snake_case")]
+enum Command {
+    #[command(description = "Setup user account")]
+    Start,
+}
+
+#[derive(Clone, Default)]
+enum GlobalState {
+    #[default]
+    Unconfigured,
+    UserSetup(UserSetup),
+    Idle,
+}
+
+#[derive(Clone)]
+enum UserSetup {
+    ReceiveFullName,
+    ReceiveAge { full_name: String },
+}
+
+/// Helper struct to store only required information to answer messages and
+/// reduce the size of the stack frames for handler functions
+#[derive(Clone)]
+struct IdsBundle {
+    chat_id: ChatId,
+    // Can be used to reply to messages or edit messages
+    // message_id: MessageId,
+}
+
+#[tokio::main]
+async fn main() {
+    pretty_env_logger::init();
+    log::info!("Starting the \"composite_state\" example");
+
+    let bot = Bot::from_env();
+
+    Dispatcher::builder(bot, schema())
+        .dependencies(dptree::deps![InMemStorage::<GlobalState>::new()])
+        .build()
+        .dispatch()
+        .await
+}
+
+fn schema() -> UpdateHandler {
+    Update::filter_message()
+        /*
+           Currently the size of the `Message` struct (for TBA 6.9) is 1936 bytes, it's insane to copy it entirely in every handler's stack.
```

Comment:
This comment looks like a discussion of `teloxide`'s tricks and implementation aspects. Don't we use plain `Message` in all the other examples?